File: CodingDojo/games/snake/src/main/dotNet/Snake/Game.Api/Board.cs
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2016 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Game.Api
{
    public class Board
    {
        public Board(String boardString)
        {
            BoardString = boardString.Replace("\n", "");
        }

        public String BoardString { get; private set; }

        /// <summary>
        /// GameBoard size (actual board size is Size x Size cells)
        /// </summary>
        public int Size
        {
            get
            {
                return (int)Math.Sqrt(BoardString.Length);
            }
        }

		public List<BoardPoint> Apples
		{
			get
			{
				return FindAllElements(BoardElement.GoodApple).ToList();
			}
		}

        /// <summary>
        /// if head is null you already died=(
        /// just skip step
        /// </summary>
		public BoardPoint? Head
		{           
			get
			{
                List<BoardPoint> result = FindAllElements(BoardElement.HeadUp)
                    .Concat(FindAllElements(BoardElement.HeadDown))
                    .Concat(FindAllElements(BoardElement.HeadLeft))
                    .Concat(FindAllElements(BoardElement.HeadRight))
                    .ToList();
				if(result.Count == 0)
                {
                    return null;
                }
                return result.Single();

			}
		}

		public List<BoardPoint> Snake
		{
			get
			{
				BoardPoint? head = Head;
				if (head == null)
				{
					return new List<BoardPoint>();
				}
				List<BoardPoint> result = FindAllElements(BoardElement.TailEndDown)
					.Concat(FindAllElements(BoardElement.TailEndLeft))
					.Concat(FindAllElements(BoardElement.TailEndRight))
					.Concat(FindAllElements(BoardElement.TailEndUp))
					.Concat(FindAllElements(BoardElement.TailHorizontal))
					.Concat(FindAllElements(BoardElement.TailVertical))
					.Concat(FindAllElements(BoardElement.TailLeftDown))
					.Concat(FindAllElements(BoardElement.TailRightDown))
					.Concat(FindAllElements(BoardElement.TailRightUp))
					.ToList();

                result.Insert(0, head.GetValueOrDefault());
				return result;
			}
		}

		public List<BoardPoint> Stones
		{
			get
			{
				return FindAllElements(BoardElement.BadApple)
					.ToList();
			}
		}

		public List<BoardPoint> Walls
		{
			get
			{
				return FindAllElements(BoardElement.Break)
					.ToList();
			}
		}

		public Direction GetSnakeDirection()
		{
			BoardPoint? head = Head;
			if (head == null)
			{
				return Direction.Left;
			}
			if (HasElementAt(head.GetValueOrDefault(), BoardElement.HeadUp))
			{
				return Direction.Up;
			}
			else if (HasElementAt(head.GetValueOrDefault(), BoardElement.HeadDown))
			{
				return Direction.Down;
			}
			else if (HasElementAt(head.GetValueOrDefault(), BoardElement.HeadRight))
			{
				return Direction.Right;
			}

			return Direction.Left;
		}

		public List<BoardPoint> GetBarriers()
		{
			return Snake
					.Concat(Stones)
					.Concat(Walls)
					.ToList();
		}


		public override string ToString() 
		{
            return String.Format("Board:{0}{1}{0}" +
                "Apple at: {2}{0}" +
                "Stones at: {3}{0}" +
                "Head at: {4}{0}" +
                "Snake at: {5}{0}" +
                "Barriers at: {6}{0}" +
                "Current direction: {7}",
                    System.Environment.NewLine,
                    PrintBoard(),
                    PrintList(Apples),
                    PrintList(Stones),
                    Head.ToString(),
                    PrintList(Snake),
                    PrintList(GetBarriers()),
                    GetSnakeDirection());
		}


		//default enginie methods------------------------------------------------

		public bool HasElementAt(BoardPoint point, BoardElement element)
        {
            if (point.IsOutOfBoard(Size))
            {
                return false;
            }

            return GetElementAt(point) == element;
        }

        public BoardElement GetElementAt(BoardPoint point)
        {
            return (BoardElement)BoardString[GetShiftByPoint(point)];
        }

        /// <summary>
        /// Writes board view to the console window
        /// </summary>
        public String PrintBoard()
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < Size; i++)
            {
                sb.Append(BoardString.Substring(i * Size, Size));
                sb.Append(System.Environment.NewLine);
            }
            return sb.ToString();
        }

        public String PrintList<T>(List<T> list)
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("[");
            foreach (var item in list)
            {
                sb.Append(item.ToString());
                sb.Append(",");
            }
            if (list.Count > 0)
            {
                sb.Remove(sb.Length - 1, 1);
            }
            sb.Append("]");
            return sb.ToString();
        }

        public List<BoardPoint> FindAllElements(BoardElement element)
        {
            List<BoardPoint> result = new List<BoardPoint>();

            for (int i = 0; i < Size * Size; i++)
            {
                BoardPoint pt = GetPointByShift(i);

                if (HasElementAt(pt, element))
                {
                    result.Add(pt);
                }
            }

            return result;
        }
		
        public bool HasElementAt(BoardPoint point, params BoardElement[] elements)
        {
            return elements.Any(elem => HasElementAt(point, elem));
        }

        public bool IsNearToElement(BoardPoint point, BoardElement element)
        {
            if (point.IsOutOfBoard(Size))
                return false;

            return HasElementAt(point.ShiftBottom(), element)
                   || HasElementAt(point.ShiftTop(), element)
                   || HasElementAt(point.ShiftLeft(), element)
                   || HasElementAt(point.ShiftRight(), element);
        }

        public bool HasBarrierAt(BoardPoint point)
        {
            return GetBarriers().Contains(point);
        }

        public int GetCountElementsNearToPoint(BoardPoint point, BoardElement element)
        {
            if (point.IsOutOfBoard(Size))
                return 0;

            //GetHashCode() in classic MS.NET for bool returns 1 for true and 0 for false;
            return HasElementAt(point.ShiftLeft(), element).GetHashCode() +
                   HasElementAt(point.ShiftRight(), element).GetHashCode() +
                   HasElementAt(point.ShiftTop(), element).GetHashCode() +
                   HasElementAt(point.ShiftBottom(), element).GetHashCode();
        }

        private int GetShiftByPoint(BoardPoint point)
        {
            return point.Y * Size + point.X;
        }

        private BoardPoint GetPointByShift(int shift)
        {
            return new BoardPoint(shift % Size, shift / Size);
        }


    }
}

File: CodingDojo/games/snake/src/main/php/Snake/api/Board.php
<?php

/**
 * Description of Board
 *
 * @author Kirill_Korol
 */

namespace Api;

class Board {

    /**
     * @var string
     */
    private $boardString;

    function __construct(string $boardString) {
        $this->boardString = str_replace("\n", "", $boardString);
    }

    public function getSize(): int {
        return sqrt(mb_strlen($this->boardString, "UTF-8"));
    }

    /**
     * @return \Api\BoardPoint[]
     */
    public function findAllElements(BoardElements $element): array {
        $result = array();
        $size = $this->getSize();
        for ($i = 0; $i < $size * $size; $i++) {
            $boardPoint = $this->getPointByShift($i);

            if ($this->hasElementAt($boardPoint, $element)) {
                array_push($result, $boardPoint);
            }
        }
        return $result;
    }

    /**
     * @param \Api\BoardPoint $point
     * @param \Api\BoardElements[] $elements
     */
    public function hasElemenstAt(BoardPoint $point, array $elements): bool {
        for ($i = 0; $i < count($elements); $i++) {
            if ($this->hasElementAt($point, $elements[$i])) {
                return true;
            }
        }
        return false;
    }

    public function hasElementAt(BoardPoint $point, BoardElements $element): bool {
        if ($point->isOutOfBoard($this->getSize())) {
            return false;
        }
        return $this->getElementAt($point) == $element;
    }

    public function getElementAt(BoardPoint $point): BoardElements {
        return BoardElements::valueOf(
                        mb_substr($this->boardString, $this->getShiftByPoint($point), 1, "UTF-8"));
    }

    public function isNearToElement(BoardPoint $point, BoardElements $element): bool {
        if ($point->isOutOfBoard($this->getSize())) {
            return false;
        }

        return $this->hasElementAt($point->shiftBottom(), $element) || $this->hasElementAt($point->shiftTop(), $element) || $this->hasElementAt($point->shiftLeft(), $element) || $this->hasElementAt($point->shiftRight(), $element);
    }

    public function getCountElementsNearToPoint(BoardPoint $point, BoardElements $element): int {
        if ($point->isOutOfBoard($this->getSize())) {
            return false;
        }

        return $this->hasElementAt($point->shiftBottom(), $element) + $this->hasElementAt($point->shiftTop(), $element) + $this->hasElementAt($point->shiftLeft(), $element) + $this->hasElementAt($point->shiftRight(), $element);
    }

    public function getShiftByPoint(BoardPoint $point): int {
        return $point->getY() * $this->getSize() + $point->getX();
    }

    public function getPointByShift(int $shift): BoardPoint {
        $size = $this->getSize();
        return new BoardPoint($shift % $size, $shift / $size);
    }

    public function printBoard(): string {
        $result = "";

        $size = $this->getSize();
        for ($i = 0; $i < $size; $i++) {
            $result .= mb_substr($this->boardString, $i * $size, $size, "UTF-8");
            $result .= "\n";
        }
        return $result;
    }

    public function hasBarrierAt(BoardPoint $point): bool {
        return in_array($point, $this->getBarriers());
    }

    /** ********************************************************************* */

    /**
     * @return \Api\BoardPoint[]
     */
    public function getApples(): array {
        return $this->findAllElements(BoardElements::$GOOD_APPLE);
    }

    /**
     * @return \Api\BoardPoint or null 
     * if head is null than you already died =( just skip step
     */
    public function getHead() {
        $result = array();
        $result = array_merge($result, $this->findAllElements(BoardElements::$HEAD_DOWN));
        $result = array_merge($result, $this->findAllElements(BoardElements::$HEAD_LEFT));
        $result = array_merge($result, $this->findAllElements(BoardElements::$HEAD_RIGHT));
        $result = array_merge($result, $this->findAllElements(BoardElements::$HEAD_UP));
        if (count($result) == 0) {
            return null;
        }
        return $result[0];
    }

    /**
     * @return \Api\BoardPoint[] or empty array if head is null
     */
    public function getSnake() {
        $head = $this->getHead();
        if ($head == null) {
            return array();
        }

        $result = array();
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_END_DOWN));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_END_LEFT));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_END_UP));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_END_RIGHT));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_HORIZONTAL));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_VERTICAL));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_LEFT_DOWN));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_LEFT_UP));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_RIGHT_DOWN));
        $result = array_merge($result, $this->findAllElements(BoardElements::$TAIL_RIGHT_UP));

        array_unshift($result, $head);
        return $result;
    }

    /**
     * @return \Api\BoardPoint[]
     */
    public function getStones(): array {
        return $this->findAllElements(BoardElements::$BAD_APPLE);
    }

    /**
     * @return \Api\BoardPoint[]
     */
    public function getWalls(): array {
        return $this->findAllElements(BoardElements::$BREAK);
    }

    public function getSnakeDirection(): Direction {
        $head = $this->getHead();
        if ($head == null) {
            return Direction::$LEFT;
        }

        if ($this->hasElementAt($head, BoardElements::$HEAD_UP)) {
            return Direction::$UP;
        } else if ($this->hasElementAt($head, BoardElements::$HEAD_DOWN)) {
            return Direction::$DOWN;
        } else if ($this->hasElementAt($head, BoardElements::$HEAD_RIGHT)) {
            return Direction::$RIGHT;
        }
        return Direction::$LEFT;
    }

    /**
     * @return \Api\BoardPoint[]
     */
    public function getBarriers(): array {

        $result = array();
        array_push($result, $this->getSnake());
        array_push($result, $this->getStones());
        array_push($result, $this->getWalls());

        return $result;
    }

    public function __toString(): string {
        return sprintf("Board:\n%s\n".
                "Apple at: %s\n" .
                "Stones at: %s\n" .
                "Head at: %s\n" .
                "Snake at: %s\n" .
                "Current direction: %s", 
                $this->printBoard(), 
                var_export($this->getApples(), true), 
                var_export($this->getStones(), true), 
                $this->getHead(), 
                var_export($this->getSnake(), true),
                $this->getSnakeDirection());
    }

}
